File: Lab2/app/src/main/java/com/example/lab2/MainActivity.java
package com.example.lab2;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    TextView input, signBox;

    String sign, value1, value2;
    Double num1, num2, result;
    boolean hasDot;
    boolean lastOp;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        input = (TextView) findViewById(R.id.input);
        signBox = (TextView) findViewById(R.id.sign);

        hasDot = false;
        lastOp = false;

        if(savedInstanceState != null){
            String firstVal, val1, val2, op;
            boolean dot;
            input.setText(savedInstanceState.getString("firstVal"));
            value1 = savedInstanceState.getString("val1");
            value2 = savedInstanceState.getString("val2");
            sign = savedInstanceState.getString("op");
            hasDot = savedInstanceState.getBoolean("dot");
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString("firstVal", input.getText().toString());
        outState.putString("op", sign);
        outState.putString("val1", value1);
        outState.putString("val2", value2);
        outState.putBoolean("dot", hasDot);
    }

    @Override
    public void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        savedInstanceState.getString("FirstVal");
        savedInstanceState.getString("val11");
        savedInstanceState.getBoolean("dot");
        savedInstanceState.getString("val2");
        savedInstanceState.getString("op");
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_0(View view) {
        String s = input.getText().toString();
        if(s.equals("0")){
            input.setText("0");
        }else{
            signBox.setText(null);
            input.setText(input.getText() + "0");
        }
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_1(View view) {
        String s = input.getText().toString();
        if(s.equals("0")){
            input.setText("1");
        }else{
            signBox.setText(null);
            input.setText(input.getText() + "1");
        }
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_2(View view) {
        String s = input.getText().toString();
        if(s.equals("0")){
            input.setText("2");
        }else{
            signBox.setText(null);
            input.setText(input.getText() + "2");
        }
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_3(View view) {
        String s = input.getText().toString();
        if(s.equals("0")){
            input.setText("3");
        }else{
            signBox.setText(null);
            input.setText(input.getText() + "3");
        }
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_4(View view) {
        String s = input.getText().toString();
        if(s.equals("0")){
            input.setText("4");
        }else{
            signBox.setText(null);
            input.setText(input.getText() + "4");
        }
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_5(View view) {
        String s = input.getText().toString();
        if(s.equals("0")){
            input.setText("5");
        }else{
            signBox.setText(null);
            input.setText(input.getText() + "5");
        }
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_6(View view) {
        String s = input.getText().toString();
        if(s.equals("0")){
            input.setText("6");
        }else{
            signBox.setText(null);
            input.setText(input.getText() + "6");
        }
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_7(View view) {
        String s = input.getText().toString();
        if(s.equals("0")){
            input.setText("7");
        }else{
            signBox.setText(null);
            input.setText(input.getText() + "7");
        }
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_8(View view) {
        String s = input.getText().toString();
        if(s.equals("0")){
            input.setText("8");
        }else{
            signBox.setText(null);
            input.setText(input.getText() + "8");
        }
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_9(View view) {
        String s = input.getText().toString();
        if(s.equals("0")){
            input.setText("9");
        }else{
            signBox.setText(null);
            input.setText(input.getText() + "9");
        }
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_dot(View view) {
        if (!hasDot) {
            if (input.getText().toString().equals("")) {
                input.setText("0.");
            } else {
                input.setText(input.getText() + ".");
            }
            hasDot = true;
        }
    }

    public void btnClick_root(View view) {
        if (input.getText().toString().equals("0")) {
            input.setText("0");
        } else {
            double num = Double.parseDouble(input.getText().toString());
            if (num < 0) {
                signBox.setText("Error!");
            } else {
                result = Math.sqrt(num);
                input.setText(result + "");
                result = null;
            }
        }
    }

    public void btnClick_sqr(View view) {
        sign = "sqr";
        value1 = input.getText().toString();
        hasDot = false;
        signBox.setText("^2");
        input.setText(Math.pow(Double.parseDouble(value1), 2) + "");
    }

    public void btnClick_power(View view){
        sign = "power";
        value1 = input.getText().toString();
        hasDot = false;
        signBox.setText("n^m");
        input.setText(null);
    }

    public void btnClick_add(View view) {
        if(lastOp == false) {
            sign = "+";
            value1 = input.getText().toString();
            input.setText(null);
            signBox.setText("+");
            lastOp = true;
            hasDot = false;
        }
        else {
            sign = "+";
            signBox.setText("+");
        }
    }

    public void btnClick_subtract(View view) {
        if(lastOp == false){
            sign = "-";
            value1 = input.getText().toString();
            input.setText(null);
            signBox.setText("-");
            hasDot = false;
            lastOp = true;
        }else{
            sign = "-";
            signBox.setText("-");
        }
    }

    public void btnClick_multiply(View view) {
        if(lastOp == false){
            sign = "*";
            value1 = input.getText().toString();
            input.setText(null);
            signBox.setText("×");
            hasDot = false;
            lastOp = true;
        }else{
            sign = "*";
            signBox.setText("×");
        }
    }

    public void btnClick_divide(View view) {
        if (lastOp == false){
            sign = "/";
            value1 = input.getText().toString();
            input.setText(null);
            signBox.setText("÷");
            hasDot = false;
            lastOp = true;
        }else{
            sign = "/";
            signBox.setText("÷");
        }
    }

    public void btnClick_factorial(View view) {
        sign = "factorial";
        hasDot = false;
        signBox.setText("!");
        value1 = input.getText().toString();
        num1 = Double.parseDouble(value1);
        double ans = 1;
        for (int i = 1; i <= num1; i ++){
            ans = ans*i;
        }
        input.setText(ans + "");
    }

    public void btnClick_percent(View view){
        sign = "percent";
        signBox.setText("%");
        hasDot = false;
        input.setText(Double.parseDouble(input.getText().toString())/100 + "");
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_nsqrt(View view){
        sign = "nsqrt";
        value1 = input.getText().toString();
        hasDot = false;
        signBox.setText("n(sqrt)");
        input.setText(null);
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_sin(View view) {
        sign = "sin";
        input.setText(null);
        hasDot = false;
        signBox.setText("sin");
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_cos(View view) {
        sign = "cos";
        input.setText(null);
        hasDot = false;
        signBox.setText("cos");
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_tan(View view) {
        sign = "tan";
        input.setText(null);
        hasDot = false;
        signBox.setText("tan");
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_log(View view) {
        sign = "log";
        input.setText(null);
        signBox.setText("log");
        hasDot = false;
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_ln(View view) {
        sign = "ln";
        input.setText(null);
        signBox.setText("ln");
        hasDot = false;
    }

    @SuppressLint("SetTextI18n")
    public void btnClick_equal(View view) {
        if (sign == null) {
            signBox.setText("Error!");
        }else if (signBox.getText().toString().equals("Error!")){
            signBox.setText(null);
            input.setText("0");
        }else if (input.getText().equals("")) {
            signBox.setText("Error!");
        } else if ((sign.equals("+") || sign.equals("-") || sign.equals("*") || sign.equals("/")) && value1.equals("")) {
            signBox.setText("Error!");
        } else {
            switch (sign) {
                default:
                    break;
                case "+":
                    value2 = input.getText().toString();
                    num1 = Double.parseDouble(value1);
                    num2 = Double.parseDouble(value2);
                    result = num1 + num2;
                    input.setText(result + "");
                    sign = null;
                    signBox.setText(null);
                    hasDot = true;
                    lastOp = false;
                    break;
                case "-":
                    value2 = input.getText().toString();
                    num1 = Double.parseDouble(value1);
                    num2 = Double.parseDouble(value2);
                    result = num1 - num2;
                    input.setText(result + "");
                    sign = null;
                    signBox.setText(null);
                    hasDot = true;
                    lastOp = false;
                    break;
                case "*":
                    value2 = input.getText().toString();
                    num1 = Double.parseDouble(value1);
                    num2 = Double.parseDouble(value2);
                    result = num1 * num2;
                    input.setText(result + "");
                    sign = null;
                    signBox.setText(null);
                    hasDot = true;
                    lastOp = false;
                    break;
                case "/":
                    value2 = input.getText().toString();
                    num1 = Double.parseDouble(value1);
                    num2 = Double.parseDouble(value2);
                    result = num1 / num2;
                    lastOp = false;
                    if(num2 == 0){
                        input.setText(null);
                        signBox.setText("Cant divide by 0!");
                        hasDot = false;
                    }else{
                        input.setText(result + "");
                        sign = null;
                        signBox.setText(null);
                        hasDot = true;
                    }
                    break;
                case "power":
                    if(input.getText().toString().equals("")){
                        signBox.setText("Error!");
                    }else {
                        value2 = input.getText().toString();
                        num1 = Double.parseDouble(value1);
                        num2 = Double.parseDouble(value2);
                        result = Math.pow(num1, num2);
                        String str2 = Double.toString(result);
                        if(str2.equals("Infinity")){
                            str2 = "0";
                        }
                        input.setText(str2);
                        signBox.setText(null);
                        lastOp = false;
                    }
                    sign = null;
                    hasDot = true;
                    break;
                case "sin":
                    value1 = input.getText().toString();
                    if(input.getText().toString().equals("")){
                        signBox.setText("Error!");
                    }else{
                        num1 = Double.parseDouble((value1));
                        input.setText(Math.sin(num1) + "");
                        signBox.setText(null);
                    }
                    sign = null;
                    hasDot = true;
                    break;
                case "cos":
                    value1 = input.getText().toString();
                    if(input.getText().toString().equals("")){
                        signBox.setText("Error!");
                    }else {
                        num1 = Double.parseDouble((value1));
                        input.setText(Math.cos(num1) + "");
                        signBox.setText(null);
                    }
                    sign = null;
                    hasDot = true;
                    break;
                case "tan":
                    if(input.getText().toString().equals("")){
                        signBox.setText("Error!");
                    }else {
                        value1 = input.getText().toString();
                        num1 = Double.parseDouble((value1));
                        input.setText(Math.tan(num1) + "");
                        signBox.setText(null);
                    }
                    sign = null;
                    hasDot = true;
                    break;
                case "log":
                    value1 = input.getText().toString();
                    if(input.getText().toString().equals("")){
                        signBox.setText("Error!");
                    }else {
                        num1 = Double.parseDouble(value1);
                        if(num1 == 0){
                            signBox.setText("Error");
                        }else{
                            input.setText(Math.log10(num1) + "");
                            signBox.setText(null);
                        }
                    }
                    sign = null;
                    hasDot = true;
                    break;
                case "ln":
                    value1 = input.getText().toString();
                    if(input.getText().toString().equals("")){
                        signBox.setText("Error!");
                    }else {
                        num1 = Double.parseDouble(value1);
                        if(num1 == 0){
                            signBox.setText("Error");
                        }else{
                            input.setText(Math.log(num1) + "");
                            signBox.setText(null);
                        }
                    }
                    sign = null;
                    hasDot = true;
                    break;
                case "nsqrt":
                    if(input.getText().toString().equals("")){
                        signBox.setText("Error!");
                    }else{
                        num1 = Double.parseDouble(value1);
                        value2 = input.getText().toString();
                        num2 = Double.parseDouble(value2);
                        input.setText(Math.round(Math.pow(num2, 1/num1)*1000.0)/1000.0 + "");
                        signBox.setText(null);
                        hasDot = true;
                    }
                    sign = null;
                    break;
            }
        }
    }
    public void btnClick_delete(View view) {
        if (input.getText().toString().equals("0")) {
            input.setText("0");
            signBox.setText(null);
        } else if(input.getText().toString().equals("")){
            input.setText("0");
        }else {
            String newStr = input.getText().toString();
            int len = input.getText().length();
            String s = newStr.substring(0, len - 1);
            input.setText(s + "");
        }
    }

    public void btnClick_clear(View view) {
        input.setText("0");
        signBox.setText(null);
        value1 = null;
        value2 = null;
        sign = null;
        hasDot = false;
        lastOp = false;
    }
}

File: onKotlin/calculator/app/src/main/java/com/example/calculator/MainActivity.kt
package com.example.calculator

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity


class MainActivity : AppCompatActivity() {

    private lateinit var input: TextView
    private lateinit var signBox: TextView

    internal var sign: String? = null
    internal var value1: String? = null
    internal var value2: String? = null
    internal var operation: String? = null
    internal var num1: Double? = null
    internal var num2: Double? = null
    internal var result: Double? = null
    internal lateinit var back: Button
    internal var hasPoint: Boolean = false
    internal var checkNumber = true
    internal var checkOperation = false
    internal var equalCliked = false


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        input = findViewById(R.id.input)
        signBox = findViewById(R.id.sign)

        back = findViewById(R.id.del)

        back.isLongClickable = true

        back.setOnLongClickListener {
            input.text = null
            signBox.text = null
            value1 = null
            value2 = null
            sign = null
            hasPoint = false
            equalCliked = false
            checkOperation = false
            true
        }
        if (savedInstanceState != null) {
            value1 = savedInstanceState.getString("number1")
            value2 = savedInstanceState.getString("number2")
            checkNumber = savedInstanceState.getBoolean("checkNumber")
            hasPoint = savedInstanceState.getBoolean("hasPoint")
            equalCliked = savedInstanceState.getBoolean("equalClicked")
            input.text = savedInstanceState.getString("numberInput")
            checkOperation = savedInstanceState.getBoolean("checkOperation")
            signBox.text = savedInstanceState.getString("signInput")
            sign = savedInstanceState.getString("sign")
            operation = savedInstanceState.getString("operation")
        }
    }

    fun btnClick_0(view: View) {
        if (equalCliked) {
            input.text = null
            if (checkNumber)
                input.text = input.text.toString() + "0"
            equalCliked = false
        } else if (checkNumber) {
            if (input.text.toString() == "0") {
                input.text = "0"
            } else
                input.text = input.text.toString() + "0"
        }
    }

    fun btnClick_1(view: View) {
        if (equalCliked) {
            input.text = null
            if (checkNumber)
                input.text = input.text.toString() + "1"
            equalCliked = false

        } else if (checkNumber) {
            if (input.text.toString() == "0")
                input.text = "1"
            else
                input.text = input.text.toString() + "1"
        }
    }

    fun btnClick_2(view: View) {
        if (equalCliked) {
            input.text = null
            if (checkNumber)
                input.text = input.text.toString() + "2"
            equalCliked = false
        } else if (checkNumber) {
            if (input.text.toString() == "0") {
                input.text = "2"
            } else
                input.text = input.text.toString() + "2"
        }

    }

    fun btnClick_3(view: View) {
        if (equalCliked) {
            input.text = null
            if (checkNumber)
                input.text = input.text.toString() + "3"
            equalCliked = false

        } else if (checkNumber) {
            if (input.text.toString() == "0") {
                input.text = "3"
            } else
                input.text = input.text.toString() + "3"
        }

    }

    fun btnClick_4(view: View) {
        if (equalCliked) {
            input.text = null
            if (checkNumber)
                input.text = input.text.toString() + "4"
            equalCliked = false

        } else if (checkNumber) {
            if (input.text.toString() == "0") {
                input.text = "4"
            } else
                input.text = input.text.toString() + "4"
        }

    }

    fun btnClick_5(view: View) {
        if (equalCliked) {
            input.text = null
            if (checkNumber)
                input.text = input.text.toString() + "5"
            equalCliked = false

        } else if (checkNumber) {
            if (input.text.toString() == "0") {
                input.text = "5"
            } else
                input.text = input.text.toString() + "5"
        }

    }

    fun btnClick_6(view: View) {
        if (equalCliked) {
            input.text = null
            if (checkNumber)
                input.text = input.text.toString() + "6"
            equalCliked = false

        } else if (checkNumber) {
            if (input.text.toString() == "0") {
                input.text = "6"
            } else
                input.text = input.text.toString() + "6"
        }

    }

    fun btnClick_7(view: View) {
        if (equalCliked) {
            input.text = null
            if (checkNumber)
                input.text = input.text.toString() + "7"
            equalCliked = false

        } else if (checkNumber) {
            if (input.text.toString() == "0") {
                input.text = "7"
            } else
                input.text = input.text.toString() + "7"
        }

    }

    fun btnClick_8(view: View) {
        if (equalCliked) {
            input.text = null
            if (checkNumber)
                input.text = input.text.toString() + "8"
            equalCliked = false

        } else if (checkNumber) {
            if (input.text.toString() == "0") {
                input.text = "8"
            } else
                input.text = input.text.toString() + "8"
        }

    }

    fun btnClick_9(view: View) {
        if (equalCliked) {
            input.text = null
            if (checkNumber)
                input.text = input.text.toString() + "9"
            equalCliked = false
        } else if (checkNumber) {
            if (input.text.toString() == "0") {
                input.text = "9"
            } else
                input.text = input.text.toString() + "9"
        }


    }

    fun btnClick_divide(view: View) {

        if (checkOperation == false) {
            sign = "/"
            value1 = input.text.toString()
            input.text = null
            signBox.text = "/"
            hasPoint = false
            checkOperation = true

        } else {
            sign = "/"
            signBox.text = "/"
        }
    }

    fun btnClick_multiply(view: View) {

        if (checkOperation == false) {
            sign = "*"
            value1 = input.text.toString()
            input.text = null
            signBox.text = "*"
            hasPoint = false
            checkOperation = true

        } else {
            sign = "*"
            signBox.text = "*"
        }

    }

    fun btnClick_subtract(view: View) {

        if (checkOperation == false) {
            sign = "-"
            value1 = input.text.toString()
            input.text = null
            signBox.text = "-"
            checkOperation = true
            hasPoint = false
        } else {
            sign = "-"
            signBox.text = "-"
        }


    }

    fun btnClick_add(view: View) {

        if (checkOperation == false) {
            sign = "+"
            value1 = input.text.toString()
            input.text = null
            signBox.text = "+"
            hasPoint = false
            checkOperation = true

        } else {
            sign = "+"
            signBox.text = "+"
        }

    }

    fun btnClick_root(view: View) {
        if(input.text.toString().equals("0")) {
            input.text = "0"
        } else {
            if(input.text.toString().toDouble() < 0 || input.text.toString().equals("")){
                signBox.text = "Error!"
            }else{
                result = Math.sqrt(input.text.toString().toDouble())
                input.text = result.toString()
            }
        }
    }

    fun btnClick_sqr(view: View) {

        if (checkOperation == false) {
            sign = "square"
            value1 = input.text.toString()
            input.text = null
            signBox.text = value1 + " ^2"
            hasPoint = false
            checkOperation = true

        } else {
            sign = "square"
            signBox.text = value1 + " ^2"
        }
    }

    fun btnClick_sin(view: View) {
        if (checkOperation == false) {
            sign = "sin"
            input.text = null
            hasPoint = false
            signBox.text = "sin"
        }
    }

    fun btnClick_cos(view: View) {
        if (checkOperation == false) {
            sign = "cos"
            input.text = null
            hasPoint = false
            signBox.text = "cos"
        }
    }

    fun btnClick_tan(view: View) {
        if (checkOperation == false) {
            sign = "tan"
            input.text = null
            hasPoint = false
            signBox.text = "tan"
        }
    }

    fun btnClick_ln(view: View) {
        if (checkOperation == false) {
            sign = "ln"
            input.text = null
            hasPoint = false
            signBox.text = "ln"
        }
    }

    fun btnClick_log(view: View) {
        if (checkOperation == false) {
            sign = "log"
            input.text = null
            hasPoint = false
            signBox.text = "log"
        }
    }

    fun btnClick_power(view: View) {
        if (checkOperation == false) {
            value1 = input.text.toString()
            sign = "npow"
            signBox.text = value1 + "^"
            input.text = null
            hasPoint = false
            checkOperation = true
        } else {
            sign = "npow"
            signBox.text = value1 + "^"

        }
    }

    fun btnClick_factorial(view: View) {
        if (checkOperation == false) {
            sign = "!"
            value1 = input.text.toString()
            hasPoint = false
            input.text = null
            checkOperation = true
            signBox.text = value1 + "!"
        } else {
            sign = "!"
            signBox.text = value1 + "!"
        }
    }

    fun btnClick_percent(view: View) {
        if (checkOperation == false) {
            sign = "%"
            value1 = input.text.toString()
            hasPoint = false
            input.text = null
            checkOperation = true
            signBox.text = "%"
        } else {
            sign = "%"
            signBox.text = "%"
        }
    }

    fun btnClick_nsqrt(view: View) {
        if (checkOperation == false) {
            sign = "nroot"
            value1 = input.text.toString()
            hasPoint = false
            input.text = null
            checkOperation = true
            signBox.text = "n based root"
        } else {
            sign = "!"
            signBox.text = "n based root"
        }
    }

    fun findFactorial(num: Int): Int {
        return if (num < 2)
            1
        else
            num * findFactorial(num - 1)
    }

    fun btnClick_equal(view: View) {
        if (sign == null) {
            signBox.text = "Error!"
        } else if (input.text == "" && sign !== "square" && sign !== "!" && sign !== "%") {
            signBox.text = "Error!"
        } else if ((sign == "+" || sign == "-" || sign == "*" || sign == "/" || sign == "square") && value1 == "") {
            signBox.text = "Error!"
        } else {
            hasPoint = true
            equalCliked = true
            when (sign) {
                "nroot" -> {
                    if(input.text.toString().equals("")) {
                        signBox.text = "Error!"
                    } else {
                        num1 = java.lang.Double.parseDouble(value1.toString())
                        value2 = input.text.toString()
                        if(input.text.toString().equals(null)){
                            input.text = "1"
                        }else{
                            num2 = java.lang.Double.parseDouble(value2.toString())
                            input.text = Math.pow(num1!!, 1 / num2!!).toString() + ""
                        }
                        signBox.text = null
                    }
                    sign = null
                    checkOperation = false
                }
                "%" -> {
                    value1 = input.text.toString()
                    if(value1.equals("")){
                        signBox.text = "Error!"
                        input.text = "0"
                    } else {
                        num1 = java.lang.Double.parseDouble(value1.toString())
                        input.text = (num1!! / 100).toString() + ""
                        sign = null
                        checkOperation = false
                        signBox.text = null
                    }
                }
                "square" -> {

                    num1 = value1?.let { java.lang.Double.parseDouble(it) }
                    result = num1!! * num1!!
                    input.text = result.toString() + ""
                    sign = null
                    signBox.text = null
                    checkOperation = false
                }
                "sin" -> {
                    value1 = input.text.toString()
                    if(value1.equals("")){
                        signBox.text = "Error!"
                        input.text = "0"
                    } else {
                        num1 = java.lang.Double.parseDouble(value1!!)
                        if (num1 == 30.0)
                            input.text = "0.5"
                        else
                            input.text = Math.sin(Math.toRadians(num1!!)).toString() + ""
                        sign = null
                        signBox.text = null
                    }
                    checkOperation = false
                }
                "cos" -> {
                    value1 = input.text.toString()
                    if(value1.equals("")){
                        signBox.text = "Error!"
                        input.text = "0"
                    } else {
                        num1 = java.lang.Double.parseDouble(value1!!)
                        if (num1 == 60.0)
                            input.text = "0.5"
                        else
                            input.text = Math.cos(Math.toRadians(num1!!)).toString() + ""
                        sign = null
                        signBox.text = null
                    }
                    checkOperation = false
                }
                "tan" -> {
                    value1 = input.text.toString()
                    if(value1.equals("")){
                        signBox.text = "Error!"
                        input.text = "0"
                    } else {
                        num1 = java.lang.Double.parseDouble(value1!!)
                        input.text = Math.tan(num1!!).toString() + ""
                        sign = null
                        signBox.text = null
                    }
                    checkOperation = false
                }
                "ln" -> {
                    value1 = input.text.toString()
                    if(input.text.toString().equals("")) {
                        signBox.text = "Error!"
                    } else {
                        if(input.text.toString().toInt() == 0) {
                            input.text = "0"
                            signBox.text = "Error!"
                        } else {
                            input.text = Math.log(value1!!.toDouble()).toString()
                            signBox.text = null
                        }
                    }
                    sign = null
                    checkOperation = false
                }
                "log" -> {
                    value1 = input.text.toString()
                    if(input.text.toString().equals("")) {
                        signBox.text = "Error!"
                    } else {
                        if(input.text.toString().toInt() == 0) {
                            input.text = "0"
                            signBox.text = "Error!"
                        } else {
                            input.text = Math.log10(value1!!.toDouble()).toString()
                            signBox.text = null
                        }
                    }
                    sign = null
                    checkOperation = false
                }
                "npow" -> {
                    if(input.text.toString().equals("")){
                        signBox.text = "Error"
                        input.text = "0"
                    } else {
                        value2 = input.text.toString()
                        num1 = java.lang.Double.parseDouble(value1.toString())
                        num2 = java.lang.Double.parseDouble(value2.toString())
                        result = Math.pow(num1!!, num2!!)
                        input.text = result!!.toString() + ""
                        checkOperation = false
                        signBox.text = null
                    }
                    sign = null
                }
                "!" -> {
                    num1 = java.lang.Double.parseDouble(value1.toString())
                    val sum = 1
                    var res = 0
                    try {
                        val n = Integer.parseInt(value1.toString())
                        res = findFactorial(n)
                        input.text = findFactorial(n).toString() + ""
                    } catch (e: NumberFormatException) {
                        input.text = value1.toString() + " is not a valid integer number"
                    }
                    checkOperation = false
                    sign = null
                    signBox.text = null
                }
                "+" -> {
                    value2 = input.text.toString()
                    num1 = java.lang.Double.parseDouble(value1.toString())
                    num2 = java.lang.Double.parseDouble(value2.toString())
                    result = num1!! + num2!!
                    input.text = result.toString() + ""
                    sign = null
                    checkOperation = false
                    signBox.text = null
                }
                "-" -> {
                    value2 = input.text.toString()
                    num1 = java.lang.Double.parseDouble(value1.toString())
                    num2 = java.lang.Double.parseDouble(value2.toString())
                    result = num1!! - num2!!
                    input.text = result.toString() + ""
                    sign = null
                    checkOperation = false

                    signBox.text = null
                }
                "*" -> {
                    value2 = input.text.toString()
                    num1 = java.lang.Double.parseDouble(value1.toString())
                    num2 = java.lang.Double.parseDouble(value2.toString())
                    result = num1!! * num2!!
                    input.text = result.toString() + ""
                    sign = null
                    checkOperation = false

                    signBox.text = null
                }
                "/" -> {
                    num1 = java.lang.Double.parseDouble(value1.toString())
                    value2 = input.text.toString()
                    num2 = java.lang.Double.parseDouble(value2.toString())
                    if (num2==0.0)
                        signBox.text = "Can't divide by zero"
                    else {
                        result = num1!! / num2!!
                        input.text = result.toString() + ""
                    }
                    checkOperation = false
                    sign = null
                }
                else -> {
                }
            }

        }
    }

    fun btnClick_dot(view: View) {
        if (!hasPoint && checkNumber) {
            if (input.text == "") {
                input.text = "0."
            } else {
                input.text = input.text.toString() + "."
            }
            hasPoint = true
        }
    }

    fun btnClick_delete(view: View) {
        if (input.text.toString().length == 1)
            input.text = "0"
        else {
            val len = input.text.length
            val s = input.text.toString()
            if (s[len - 1] == '.') {
                hasPoint = false
                input.text = input.text.subSequence(0, len - 1)
            } else
                input.text = input.text.subSequence(0, len - 1)
        }
    }

    fun btnClick_clear(view: View) {
        input.text = "0"
        signBox.text = null
        value1
        value2 = null
        sign = null
        hasPoint = false
        equalCliked = false
        checkOperation = false
    }


    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString("numberInput", input.text.toString())
        outState.putString("signInput", signBox.text.toString())
        outState.putString("sign", sign)
        outState.putString("number1", value1)
        outState.putString("number2", value2)
        outState.putBoolean("equalClicked", equalCliked)
        outState.putBoolean("checkNumber", checkNumber)
        outState.putBoolean("checkOperation", checkOperation)
        outState.putBoolean("hasPoint", hasPoint)
        operation = signBox.text.toString()
        outState.putString("operation", operation)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        savedInstanceState.getString("numberInput")
        savedInstanceState.getString("number1")
        savedInstanceState.getString("number2")
        savedInstanceState.getString("checkNumber")
        savedInstanceState.getString("hasPoint")
        savedInstanceState.getString("signInput")
        savedInstanceState.getString("sign")
        savedInstanceState.getString("equalClicked")
        savedInstanceState.getString("checkOperation")
        savedInstanceState.getString("operation")

    }


}
